pkg/ratelimit/grpc.go

```diff
@@ -0,0 +1,125 @@
+// Copyright © 2021 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ratelimit
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/auth"
+	"go.thethings.network/lorawan-stack/v3/pkg/events"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+	"go.thethings.network/lorawan-stack/v3/pkg/rpcmetadata"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/peer"
+)
+
+// GRPCVaryBy returns a rate limit key from a request context and the full method name.
+type GRPCVaryBy func(ctx context.Context, fullMethod string) string
+
+// GRPCRemoteIP returns the remote IP of a gRPC request.
+func GRPCRemoteIP(ctx context.Context, fullMethod string) string {
+	if md := rpcmetadata.FromIncomingContext(ctx); md.XForwardedFor != "" {
+		xff := strings.Split(md.XForwardedFor, ",")
+		return strings.Trim(xff[0], " ")
+	}
```

Comment:
We have the `ProxyHeaders` middleware that sets `x-real-ip` when a trusted proxy sets those headers (and otherwise unsets it)